src/Executor/Promise/Adapter/AmpPromiseAdapter.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+declare(strict_types=1);
+
+namespace GraphQL\Executor\Promise\Adapter;
+
+use Amp\Deferred;
+use Amp\Failure;
+use Amp\Promise as AmpPromise;
+use Amp\Success;
+use GraphQL\Executor\Promise\Promise;
+use GraphQL\Executor\Promise\PromiseAdapter;
+use Throwable;
+use function Amp\Promise\all;
+use function array_replace;
+
+class AmpPromiseAdapter implements PromiseAdapter
+{
+    /**
+     * @inheritdoc
+     */
+    public function isThenable($value) : bool
+    {
+        return $value instanceof AmpPromise;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function convertThenable($thenable) : Promise
+    {
+        return new Promise($thenable, $this);
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function then(Promise $promise, ?callable $onFulfilled = null, ?callable $onRejected = null) : Promise
+    {
+        $deferred = new Deferred();
+
+        // @see https://promisesaplus.com/#point-41
+        $onResolve = static function (?Throwable $reason, $value) use ($onFulfilled, $onRejected, $deferred) : void {
+            if ($reason === null && $onFulfilled === null) {
+                $deferred->resolve($value);
+            } elseif ($reason === null) {
+                self::resolveWithCallable($deferred, $onFulfilled, $value);
+            } elseif ($onRejected === null) {
+                $deferred->fail($reason);
+            } else {
+                self::resolveWithCallable($deferred, $onRejected, $reason);
+            }
+        };
+
+        /** @var AmpPromise $adoptedPromise */
+        $adoptedPromise = $promise->adoptedPromise;
+        $adoptedPromise->onResolve($onResolve);
+
+        return new Promise($deferred->promise(), $this);
```

Comment:
I am curious why do you need this method at all? Why not just do:

```php
public function then(Promise $promise, ?callable $onFulfilled = null, ?callable $onRejected = null) : Promise
{
        /** @var AmpPromise $adoptedPromise */
        $adoptedPromise = $promise->adoptedPromise;
        return new Promise($adoptedPromise->then($onFulfilled, $onRejected), $this); 
}
```

Reply:
Amphp doesn't have a `then`. It has `onResolve` instead but it only takes one callable and is called regardless of whether the promise was resolved with a success or with a failure.

See https://github.com/amphp/amp/blob/13930a582947831bb66ff1aeac28672fd91c38ea/lib/Promise.php#L24-L25

Reply:
Ah, I see. I didn't realize they are not conforming to A+ spec and just have their own implementation

Reply:
Yup. Amphp was created to use coroutines and avoid callbacks. Apparently Promises A+ are not suitable for that:

```
Amp offers a single callback registration mechanism, onResolve(), which is optimized for coroutines. We think that coroutines are superior to callbacks.
```

Source: https://github.com/amphp/amp/issues/150#issuecomment-322245990